core/domain/topic_domain.py

```diff
@@ -75,12 +83,31 @@ def __init__(self, change_dict):
             raise Exception('Invalid change_dict: %s' % change_dict)
         self.cmd = change_dict['cmd']
 
-        if self.cmd == CMD_UPDATE_TOPIC_PROPERTY:
+        if self.cmd == CMD_ADD_SUBTOPIC:
+            self.id = change_dict['subtopic_id']
+            self.title = change_dict['title']
+        elif self.cmd == CMD_DELETE_SUBTOPIC:
+            self.id = change_dict['subtopic_id']
+        elif self.cmd == CMD_ADD_UNCATEGORIZED_SKILL_ID:
+            self.id = change_dict['new_uncategorized_skill_id']
+        elif self.cmd == CMD_REMOVE_UNCATEGORIZED_SKILL_ID:
+            self.id = change_dict['uncategorized_skill_id']
+        elif self.cmd == CMD_MOVE_SKILL_ID_TO_SUBTOPIC:
+            self.id = change_dict['subtopic_id']
+            self.skill_id = change_dict['skill_id']
```

Comment:
You know, sorry about this, but I just realized that in terms of reverting it might be a good idea after all to have old_subtopic_id and new_subtopic_id as stuff passed in (in addition to skill_id). Either of those two could be None (meaning uncategorized), and then the actual move method would just do validation directly. The advantage of recording the old subtopic ID is that you can undo the change if needed.

I also think self.new_subtopic_id is a better name than self.id, because the latter is ambiguous since there are multiple objects involved here.

Reply:
Done, also changed the command to CMD_MOVE_SKILL_ID as the destination can be uncategorized_skill_ids also.